### PR TITLE
Fix BuildConfig SourceSecret Sync

### PR DESCRIPTION
**Pull request by waveywaves**
Fixes https://github.com/openshift/jenkins-sync-plugin/issues/359 https://github.com/openshift/jenkins-sync-plugin/issues/355
- Added Logging
- Little bit of Refactor
- upsertCredentials returns new credId which was breaking the feature

### CODE REVIEW

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 1]
```diff
@@ -63,7 +64,9 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
 		BuildSource source = null;
 		String jenkinsfile = null;
 		String jenkinsfilePath = null;
+    GitSCM gitSCM = null;
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: this variable is not used anywhere else than in https://github.com/openshift/jenkins-sync-plugin/pull/360/files#diff-fbc124cdb4f21738ce080e4247a0a6fbR89 

We should declare it only there.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 2]
```diff
@@ -63,7 +64,9 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
 		if (spec != null) {
+		  LOGGER.info("Mapping BuildConfig "+bc.getMetadata().getName()+" to FlowDefinition");
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: Requires:
- proper formatting for String concatenation (add spaces)
- null pointer protection in the log message 

if this cannot be performed, better to just display bc.getMetadata() and check bc != null

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 3]
```diff
@@ -106,6 +98,31 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
 		}
 	}
 
+	public static GitSCM getGitSCM(BuildSource source, BuildConfig bc){
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: does it require to be public? and static ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 4]
```diff
@@ -106,6 +98,31 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+    GitBuildSource gitSource = source.getGit();
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: we need null safety check

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 5]
```diff
@@ -106,6 +98,31 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+    String branchRef = gitSource.getRef();
+    List<BranchSpec> branchSpecs = Collections.emptyList();
+    GitSCM gitSCM = null;
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: do we need declaration here?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 6]
```diff
@@ -106,6 +98,31 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+    String credentialsId = null;
+    LocalObjectReference sourceSecret = source.getSourceSecret();
+    if (sourceSecret != null){
+      try{
+        credentialsId = updateSourceCredentials(bc);
+      } catch (Exception e){
+        LOGGER.warning("Encountered "+e.toString()+" during updateSourceCredentials()");
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: call to toString is implicit, no need to specify it here.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 7]
```diff
@@ -106,6 +98,31 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+      }
+    }
+    if (isNotBlank(branchRef)) {
+      branchSpecs = Collections.singletonList(new BranchSpec(branchRef));
+    }
+    // if credentialsID is null, go with an SCM where anonymous has to be sufficient
+    gitSCM = new GitSCM(
+        Collections.singletonList(new UserRemoteConfig(gitSource.getUri(), null, gitSource.getRef(), credentialsId)),
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: gitSource.getRef() has already been set in a variable previously. Better to you the declared variable.

Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: Add an intermediate local variable or even 2.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 8]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
-			CpsScmFlowDefinition cpsScmFlowDefinition = (CpsScmFlowDefinition) definition;
-			String scriptPath = cpsScmFlowDefinition.getScriptPath();
-			if (scriptPath != null && scriptPath.trim().length() > 0) {
-				boolean rc = false;
-				BuildSource source = getOrCreateBuildSource(spec);
-				String bcContextDir = source.getContextDir();
-				if (StringUtils.isNotBlank(bcContextDir) && scriptPath.startsWith(bcContextDir)) {
-					scriptPath = scriptPath.replaceFirst("^" + bcContextDir + "/?", "");
-				}
+    LOGGER.info("Updating BuildConfig From Job "+namespaceName);
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: add spaces between the "+" sign

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 9]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
 
-				if (!scriptPath.equals(jenkinsPipelineStrategy.getJenkinsfilePath())) {
-					LOGGER.log(Level.FINE,
-							"updating bc " + namespaceName + " jenkinsfile path to " + scriptPath + " from ");
-					rc = true;
-					jenkinsPipelineStrategy.setJenkinsfilePath(scriptPath);
-				}
+    FlowDefinition definition = job.getDefinition();
+    if (definition instanceof CpsScmFlowDefinition) {
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: null check ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 10]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+      BuildSource source = getOrCreateBuildSource(spec);
+      CpsScmFlowDefinition cpsScmFlowDefinition = (CpsScmFlowDefinition) definition;
+      String scriptPath = cpsScmFlowDefinition.getScriptPath();
+      SCM scm = cpsScmFlowDefinition.getScm();
+      if (scm instanceof GitSCM){
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: null check ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 11]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+        GitSCM gitSCM = (GitSCM) scm;
+        List<UserRemoteConfig> userRemoteConfigs = gitSCM.getUserRemoteConfigs();
+        LocalObjectReference sourceSecret = source.getSourceSecret();
+        if  (sourceSecret != null){
+          UserRemoteConfig sourceSecretUserRemoteConfig = new UserRemoteConfig(null, null,null,sourceSecret.getName());
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: add local variable for sourceSecret.getName()

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 12]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+          if (!userRemoteConfigs.contains(sourceSecretUserRemoteConfig)){
+            LOGGER.info("Adding Build SourceSecret "+sourceSecret.getName()+" as UserRemoteConfig");
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: spaces for String concat

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 13]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
 	 * @return true if the BuildConfig was changed
 	 */
 	public static boolean updateBuildConfigFromJob(WorkflowJob job, BuildConfig buildConfig) {
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: we need additional comments for this method especially as it contains many return statements

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java [hunk 14]
```diff
@@ -116,88 +133,103 @@ public static FlowDefinition mapBuildConfigToFlow(BuildConfig bc) throws IOExcep
...
+      return false;
+    }
 
-			return false;
-		}
+    if (definition instanceof CpsFlowDefinition) {
+      CpsFlowDefinition cpsFlowDefinition = (CpsFlowDefinition) definition;
+      String jenkinsfile = cpsFlowDefinition.getScript();
+      if (jenkinsfile != null && jenkinsfile.trim().length() > 0
+        && !jenkinsfile.equals(jenkinsPipelineStrategy.getJenkinsfile())) {
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: add local variable

File: src/main/java/io/fabric8/jenkins/openshiftsync/CredentialsUtils.java [hunk 15]
```diff
@@ -51,37 +54,58 @@
...
-            Secret sourceSecret = getAuthenticatedOpenShiftClient().secrets()
-                    .inNamespace(buildConfig.getMetadata().getNamespace())
-                    .withName(buildConfig.getSpec().getSource().getSourceSecret().getName()).get();
-            return sourceSecret;
+    BuildConfigSpec spec = buildConfig.getSpec();
+    if (spec != null) {
+      BuildSource source = spec.getSource();
+      if (source != null) {
+        LocalObjectReference sourceSecret = source.getSourceSecret();
+        if (sourceSecret != null && !sourceSecret.getName().isEmpty()) {
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/CredentialsUtils.java]: null check on sourceSecret.getName() != null

File: src/main/java/io/fabric8/jenkins/openshiftsync/CredentialsUtils.java [hunk 16]
```diff
@@ -51,37 +54,58 @@
...
+          ObjectMeta buildConfigMetadata = buildConfig.getMetadata();
+          String namespace = buildConfigMetadata.getNamespace();
+          String buildConfigName = buildConfigMetadata.getName();
+          String sourceSecretName = sourceSecret.getName();
+          logger.info("Retrieving SourceSecret for BuildConfig "+buildConfigName+" in Namespace "+namespace);
+          Secret secret = getAuthenticatedOpenShiftClient().secrets().inNamespace(namespace).withName(sourceSecretName).get();
+          if (secret != null){
+            return secret;
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/CredentialsUtils.java]: testing nullity rather non-nullity would improve readability.

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java (hunk 3 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: I ahve changed it to private, it needs to be static though, as we are referencing it from a static context,

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java (hunk 9 above)
Comment (waveywaves) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildConfigToJobMapper.java]: instanceof will be false if `definition` is `null`.

**Pull request merged**